### DISCUSSION TITLE
Add `intl` extension

### DISCUSSION THREAD

**Discussion by hellomohsinali** [NativePHP/laravel#292] [Ideas]
```
function format_number(float | int $number): string
 {

        $formatter = new NumberFormatter(app()->getLocale(), NumberFormatter::DECIMAL);

        return $formatter->format($number);
}

format_number(2);
```

This code won't run as expected.

**Comment by MarceauKa** [comment]
Fixed it requiring `symfony/polyfill-intl-icu`

**Comment by buchin** [comment]
> Fixed it requiring `symfony/polyfill-intl-icu`

Hi @MarceauKa, could you please elaborate required step to solve this issue?
I got this issue and requiring symfony/polyfill-intl-icu does not resolve it.

Thank you

**Comment by pmffromspace** [comment]
Hi I did a lot of research and this is the closest I got to an solution for this exact problem. Im trying to pack a filamenphp laravel webapp into a desktop programm with the help of nativePHP and I just cant find a way to remove the error „php extension intl not found“

@MarceauKa could you please just quickly tell us in which file (and maybe how) you fixed this?
I tried a lot of ways already, but Im unfortunatly new to php so I would really appreciate the help! :)

Kind regards

**Comment by simonhamp** [comment]
This is because the `intl` extension isn't built into the PHP binaries we're using yet.

I also noticed that it's [apparently not available for Windows](https://static-php.dev/en/guide/extensions.html) yet.

I will test and see what needs doing to get this in place on the binary side so we don't need to mess around with polyfills.

**Comment by pmffromspace** [comment]
Thanks for this quick response!

In the meantime I tried to build the php binaries manually with this approach:

```
composer require nativephp/php-bin --dev

git clone https://github.com/crazywhalecc/static-php-cli.git --depth=1
cd static-php-cli
bin/spc download --clean
bin/spc download --with-php=8.2 --for-extensions "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"
bin/spc build --build-cli --build-embed "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"

##-> copying to vendor directory...
# Doing the same for 8.3

bin/spc download --clean
bin/spc download --with-php=8.3 --for-extensions "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"
bin/spc build --build-cli --build-embed "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"

```
I then (Im on MacOS) right clicked the php bin and compressed it to a .zip file and renamed and replaced the supposingly old php bins at `vendor/nativephp/php-bin/bin/mac/arm64/php-8.2.zip`

But I stillt get the same error. (with `php artisan native:serve`or `php artisan native:build`)

Did I miss something when replacing the php binary or is it just not that simple?

Kind regards.

**Comment by simonhamp** [comment]
In theory, what you've done is correct...

What version of PHP are you running locally? `php -v`

**Comment by pmffromspace** [comment]
`PHP 8.3.6 (cli) (built: Apr 12 2024 00:13:56) (NTS)`

maybe I should zip it using a cli tool?

**Comment by simonhamp** [comment]
So the thing to check is if the PHP you've built actually includes `intl`

So you can check the PHP8.3 binary you generated:

```
/path/to/binary/php -i | grep intl
```

**Comment by pmffromspace** [comment]
I get following (I cant interpret if this is sufficient)

`cd <directroy-of-generated-php-bin>`
`./php-8.3 -i | grep intl`

Output:
```
intl
intl.default_locale => no value => no value
intl.error_level => 0 => 0
intl.use_exceptions => Off => Off
```

**Comment by simonhamp** [comment]
That means intl extension is enabled and working

**Comment by simonhamp** [comment]
It's possible that even after copying it into the vendor folder it gets wiped away

Can you also check the version that gets built into the app?

**Comment by pmffromspace** [comment]
Im having trouble finding a way to check that - in the developer tools I cant find the php version. Im also unable to build php-8.1.
How would I check the final build php version?

**Comment by simonhamp** [comment]
There are some [instructions in the docs](https://nativephp.com/docs/getting-started/debugging#check-your-app-and-php)

**Comment by pmffromspace** [comment]
I was able to **solve this issue** 🚀with following steps thanks to the very helpful and kind person @simonhamp - many thanks!! (I will list all above discussed steps again so people stumbeling over it with the same issue have a better overview:

---
## DISCLAIMER:

1. !! adapt the manual build to your php-version !!
2. I think this work around only works for the OS+Arch your manually building and replacing the php bin!

----

### Setup:
- Filamentphp
- Laravel
- nativePHP
- basically the app enables users to input data into tables with the help of forms and displays them with charts on the app dashboard
-> Trying to create a desktop app version off my laravel filament web app.

### Problem:
- everything with nativePHP works fine, Desktop App starts and looks good. 
- starting desktop app I sometimes got `error code 500`
- When user creates a entry (fills a form to add row to table) there is a fatal php error "php extension intl not found"

### Cause:
the static php binaries currently used by nativePHP dont include this extension (yet)

---

# Solution:
- basically I got it working by manually building the php binary and replacing the one used by nativePHP.

## How I did this:

## 1. building php binary
- clone git for static-php git
- !! adapt to your php-version !!
```
composer require nativephp/php-bin --dev

git clone https://github.com/crazywhalecc/static-php-cli.git --depth=1
cd static-php-cli
bin/spc download --clean
bin/spc download --with-php=8.3 --for-extensions "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"
bin/spc build --build-cli --build-embed "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"

```

## 2. move php binary to nativePHP

- rename original `/path/to/your/app/vendor/nativephp/electron/resources/js/resources/php/php`
to `php.original`

copy the newly built php binary from `/static-php-cli/buildroot/bin/php` to the nativePHP directory of your project `/path/to/your/app/vendor/nativephp/electron/resources/js/resources/php/php`


## 3. build with nativePHP
- run `php artisan native:build` and/or `php artisan native:serve`

---

_i: Interestingly, it seems like the php binary gets copied from the dev to the prod build at dist so it seems to work._

### verifying the build worked:

#### checking newly build php bin:
```
cd /static-php-cli/buildroot/bin/
./php -i | grep intl
./php -v
```


#### checking after 'php artisan native:build` run
```
# for DEV build
cd /path/to/your/app/vendor/nativephp/electron/resources/js/resources/php/php
./php -i | grep intl
./php -v

# for PROD build
cd /path/to/your/app/dist/mac-arm64/YOU_APP/Contents/Resources/app.asar.unpacked/resources/php/
./php -i | grep intl
./php -v
```

#### Helpful link to debug nativePHP related stuff: [https://nativephp.com/docs/1/getting-started/debugging#check-your-app-and-php](https://nativephp.com/docs/1/getting-started/debugging#check-your-app-and-php )

**Reply by simonhamp** [reply]
Nice write-up!

I just want to re-state that if you're using custom builds of PHP you may have issues with distribution later down the line - I'm not clear yet if these PHP builds also need signing in order to not be blocked (if anyone knows better, please chime in!)

Also, from a support perspective, it's helpful to the maintainers to have a limited range of versions of PHP to try and debug issues with. I don't want to stop you building your apps, but any custom PHP builds will require paid support and won't be supported via GitHub.

**Reply by pmffromspace** [reply]
Generally speaking: Do you plan to add the `intl` extension to the main builds? Because I think it would be needed to support "Laravel Filamentphp" developers -> major plus for whole nativePHP project. :)

**Reply by simonhamp** [reply]
Yes definitely

**Reply by simonhamp** [reply]
It's just waiting on [this](https://github.com/crazywhalecc/static-php-cli/issues/435)

**Reply by simonhamp** [reply]
So, sadly _baked-in_ Windows `intl` support isn't coming anytime soon. But it's possible that we can still load it dynamically. I'm going to look into this shortly

**Reply by mansoorkhan96** [reply]
I am planning to build a couple apps with this and hit this issue on first one. I am also using Filament. Would be nice to know if there is any update on this and if its going to be possible in future.

**Reply by simonhamp** [reply]
@mansoorkhan96 the update as of today is: statically compiled `intl` [still isn't supported on Windows](https://static-php.dev/en/guide/extensions.html) and I haven't as yet had chance to try getting a version of the Windows binaries working with it loaded dynamically

once we've sorted a way for us to have `intl` support on Windows, we can ship builds with `intl` for all platforms

if you're not targeting Windows tho, `intl` _is_ supported for macOS and Linux, so you could build your own PHP binary that has `intl` baked in for those platforms

**Reply by mansoorkhan96** [reply]
Thanks for the update Simon.
I am on mac but yes the target platform is Windows. 
I will try building the PHP binary on my own for some testing.
I will also look into creating binaries for windows too. I will drop you an update here if i can find a way.

**Reply by rogerznts** [reply]
> pmffromspace

Hi, I'm using Filament with NativePHP and I was having the same problem. Unfortunately, the solution didn't work for me. Every time I built or entered dev mode, the PHP binary was replaced. In this case, I created a PHP folder and had to create the structure. Passing the path of this PHP folder to the constant in the env. NATIVEPHP_PHP_BINARY_PATH. Also, the cert.pem file is important.

![CleanShot 2024-09-20 at 16 19 00](https://github.com/user-attachments/assets/bbad5fe1-02ba-4d21-b18b-dbe68676e11d)

This way, it keeps the correct version during the build or dev.

I'm also trying to resolve two errors that are preventing the top right corner menu from working. Among other javascript issues with file upload.

support.js?v=3.2.113.0 
TypeError: Cannot read properties of undefined (reading 'from')

and

livewire.js?id=cc800bf4:1129 
Uncaught TypeError: window.pluralize is not a function

If you've experienced this and have already resolved it, please share :)

**Reply by pmffromspace** [reply]
tbh I did not do any filament-NativePHP stuff since this post, Im not sure if I might consider different cross-platform stategies in the future to develop my small applications because I want to focus on functionality and do less debugging...
Sorry I cant help with that

**Reply by rogerznts** [reply]
I understand, I'm developing an application to solve my problem, without any great expectations. I'll continue exploring this combination, which is very interesting.

**Reply by simonhamp** [reply]
For the support.js issue, [check here](https://github.com/orgs/NativePHP/discussions/345)

**Reply by rogerznts** [reply]
@simonhamp awesome! Thanks

**Reply by kdrmlhcn** [reply]
I solved problem on MacOS but it still exists on Windows. Is anyone solved this issue on Windows? @simonhamp @pmffromspace

**Reply by danie1net0** [reply]
I'm on MacOS and reproduced these steps, but they didn't solve the problem. I copied the PHP bin from `/opt/homebrew/Cellar/php/8.4.2/bin/php` to `vendor/nativephp/electron/resources/js/resources/php/` and this is the output when I run the command `./php -i | grep intl`:
```
intl
intl.default_locale => no value => no value
intl.error_level => 0 => 0
intl.use_exceptions => Off => Off
```

But when I open the Filament listing page, the error continues.

**Comment by robot-iraq** [comment]
I was excited to discover this tool, but the PHP development experience on Windows is far from ideal. It's not just the "intl" extension that's essential for PHP development—other critical extensions like PCNTL, which is used in Horizon, Octane, and Laravel's concurrency features, are also unavailable on Windows. Unfortunately, I don’t see any support or implementation for these extensions coming anytime soon.

**Reply by danie1net0** [reply]
Actually Windows can't provide you a good PHP development experience in general.

**Reply by robot-iraq** [reply]
well, is it the OS responsibility, or the language?

**Reply by simonhamp** [reply]
`pcntl` is a very rich and robust abstraction over key process management features of *nix-like operating systems. Windows completely lacks this level of robustness by default.

So for that one, it is the OS's fault and really Microsoft's responsibility to provide better tooling there that is on par with the industry standard.

Maybe WSL2 goes some way to helping, but running applications that rely on that feels brittle at best.

The lack of `intl` is just a result of not finding the right path to compile it statically for Windows yet, and is neither a language nor OS problem - it's just a combination of things that very few people have tried to tackle.

**Comment by Mghobadid** [comment]
**Workaround to Bypass the Intl Extension Check in Laravel Number Helper (For Windows Users)**

For anyone facing this issue on Windows, I came up with a (call it "bad but working") way to bypass the intl extension check used in Laravel's Number Helper. This helper was added to Laravel starting from version 10.33.0, so if you are using that version or higher, you might not face this problem.

In this temporary workaround, we trick the PHP interpreter into assuming that the intl extension is loaded!

⚠️⚠️ **Attention:** ⚠️⚠️

This method could potentially harm your application and introduce risks, so use it only if you 100% understand what you're doing. By bypassing this check, you are essentially telling the application that the intl extension is available, even if it isn't, which can lead to unexpected behavior in other parts of the framework or packages that rely on it.

**Here's how it works:**

Create an Override for the extension_loaded function: We define a custom version of the extension_loaded function inside the Illuminate\Support namespace, which forces it to always return true when checking for the intl extension.

Load the Override Before Laravel's Core: The trick is to load this override file before Laravel’s autoloader. This ensures that whenever the Number Helper (or any other part of Laravel) checks if intl is loaded, our custom function is used.

**Test Carefully:** 
Ensure that this doesn't break any other functionality, especially parts of the framework or third-party packages that rely on the actual availability of the intl extension.

**The Override Code:**

```
<?php
// app/Overrides/intl_override.php

namespace Illuminate\Support;

if (! function_exists('Illuminate\Support\extension_loaded')) {
    function extension_loaded($name)
    {
        if ($name === 'intl') {
            return true; // Assume intl is loaded
        }
        return \extension_loaded($name); // Call the original function for others
    }
}
```
Then, in your public/index.php, include this file before Laravel's autoloader:

```
<?php
// public/index.php

// Preload your override for extension_loaded in the Illuminate\Support namespace.
require __DIR__ . '/../app/Overrides/intl_override.php';

require __DIR__ . '/../vendor/autoload.php';

// ... rest of your bootstrap process
```
**Final Thoughts:**

- **Use With Caution:** This is a hacky solution that forces Laravel to ignore the missing intl extension, but it's not a sustainable or recommended practice. Only use this if absolutely necessary, and test your application thoroughly.

- **Risk Assessment:** If Laravel or any package relies on the actual functionality of the intl extension, this could cause unpredictable behavior. 

- **Future Improvements:** I believe the final and proper workaround would be to load the intl.dll dynamically into static PHP. Unfortunately, I haven't been able to implement this solution yet, but it could potentially provide a more stable and permanent fix.

**Comment by binaryfire** [comment]
I just saw this comment in the `static-php-cli` repo: https://github.com/crazywhalecc/static-php-cli/issues/435#issuecomment-2760017586. Does that mean it's now possible to add it for Windows?

**Reply by patimio66** [reply]
@simonhamp

**Comment by nphuly** [comment]
Any chance that I can use/add this extension on mac arm64? Thanks!

**Reply by PeteBishwhip** [reply]
> Any chance that I can use/add this extension on mac arm64? Thanks!

Check out [this answer](https://github.com/orgs/NativePHP/discussions/292#discussioncomment-9314788) from @pmffromspace above. Its definitely a way to achieve this on a case-by-case basis. 😊

**Reply by nphuly** [reply]
[This answer](https://github.com/orgs/NativePHP/discussions/292#discussioncomment-9314788) doesn't resolve my issue. So here is my workaround that works for me:

### Problem
Every time you run the `php artisan native:build` or `php artisan native:serve`, `electron` will remove the `php` binary in this path `/path/to/your/app/vendor/nativephp/electron/resources/js/resources/php/php`, and replace it with the pre-built binary from `nativephp/php-bin`. If you follow step 2 of the mentioned answer, it won't work.

### Solution

#### Step 1. Build the binary
```
composer require nativephp/php-bin --dev

git clone https://github.com/crazywhalecc/static-php-cli.git --depth=1

cd static-php-cli

composer update

bin/spc download --clean

bin/spc download --with-php=8.3 --for-extensions "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"

bin/spc build --build-cli --build-embed "intl,bcmath,ctype,curl,dom,fileinfo,filter,mbstring,openssl,pdo,pdo_sqlite,session,simplexml,sockets,sqlite3,tokenizer,xml,zlib"
```

#### Step 2. Add execute perm and zip the built binary

```
cd static-php-cli/buildroot/bin

# Add execute perm (Linux and macOS only)
chmod +x php

# macOS / Linux
zip php-[PHP_MAJOR_VERSION].[PHP_MINOR_VERSION].zip php
```

#### Step 3. Create a customized `php/bin` structure and add NATIVEPHP_PHP_BINARY_PATH
Followed this instruction: https://nativephp.com/docs/desktop/1/digging-deeper/php-binaries#building-custom-binaries

In my project root folder, I created a new folder structure `php/bin` under the `resources` folder. The final structure should be like this:

```
resources
├── css
│   └── app.css
├── js
│   ├── app.js
│   └── bootstrap.js
├── php
│   └── bin
│       └── mac
│           └── arm64
└── views
    └── welcome.blade.php
```
Depending on your needs, create the corresponding OS/architecture folders. Then in my `.env`, I added this:
```
NATIVEPHP_PHP_BINARY_PATH=resources/php/
```

#### Step 4. Copy the zip file and run
Now, just copy the zipped `php` file(s) to the destination OS and architecture folder(s), mine looks like this:
```
resources
├── css
├── js
└── php
    └── bin
        └── mac
            └── arm64
                └── php-8.3.zip
```
Now, just run the `php artisan native:build` or `php artisan native:serve`.

Thank you, and hope this helps!

**Comment by PeteBishwhip** [comment]
I'm pleased to advise that support for Intl has now been made available by the static-php team. We've applied this to our latest PHP bin builds and it seems to be working.

I anticipate this being built out and released within the next couple of days. Keep an eye out for an announcement!

**Reply by patimio66** [reply]
Thanks a lot! Good news ❤️

**Reply by mansoorkhan96** [reply]
Is this working on Mac? I have tried on Mac and i still get the error.

**Reply by PeteBishwhip** [reply]
You'll need to wait until the next release, or alternatively, clone the repo PR that includes intl and use it for a custom binary build until it's released.

**Reply by mansoorkhan96** [reply]
ah i thought they have released already but they just merged the commit.

However, I tried to use the main branch which has the intl commit but I still get error.

I am using: "nativephp/php-bin": "dev-main as 1.0.2"

<img width="1145" alt="image" src="https://github.com/user-attachments/assets/5d37ad25-6127-4534-955b-c155c4583724" />

**Comment by simonhamp** [comment]
Great news all! `intl` is now available across all desktop platforms 🎉 

Make sure you have `php-bin@1.0.3+`

Thanks for your patience 🙏🏼

**Reply by patimio66** [reply]
Thank you Simon ❤️🫰 Will check it later.

**Comment by Behruz1905** [comment]
Hi everyone,

I'm encountering an issue with my PHP 8.2. I am using FilamentPHP and I'm seeing the same error in my application.

I'm specifically getting the Undefined array key error. Has anyone else experienced this, and could you offer some guidance or a potential fix?
